src/data.c

```diff
@@ -3,37 +3,37 @@
 
 // Initialised at load time
 SEXP syms_vec_proxy = NULL;
-SEXP syms_vec_proxy_dispatch = NULL;
-SEXP fns_vec_proxy_dispatch = NULL;
 
 
-static SEXP vec_proxy_dispatch(SEXP x) {
-  return vctrs_dispatch1(syms_vec_proxy_dispatch, fns_vec_proxy_dispatch,
-                         syms_x, x);
-}
-
 // [[ register(); include("vctrs.h") ]]
 SEXP vec_proxy(SEXP x) {
-  if (vec_typeof(x) == vctrs_type_s3) {
-    return vec_proxy_dispatch(x);
+  int nprot = 0;
+  struct vctrs_type_info info = PROTECT_TYPE_INFO(vec_type_info(x), &nprot);
+
+  SEXP out;
+  if (info.type == vctrs_type_s3) {
+    out = vec_proxy_invoke(x, info.proxy_method);
   } else {
-    return x;
+    out = x;
   }
+
+  UNPROTECT(nprot);
+  return out;
 }
 // [[ include("vctrs.h") ]]
 SEXP vec_proxy_method(SEXP x) {
   return s3_find_method("vec_proxy", x);
 }
 // [[ include("vctrs.h") ]]
 SEXP vec_proxy_invoke(SEXP x, SEXP method) {
```

Comment:
I think you could remove this function now? It's also used in `vctrs_proxy_info()`, but it still seems inlinable.

Reply:
I could also remove `vec_proxy_method()` but I prefer to keep both wrappers for readability. Also that would still require to forward declare `syms_vec_proxy`. I have removed both these methods from `vctrs.h` to underline that they are purely internal.